### ISSUE TITLE
Unknown types narrowed to an intersection won't narrow further

### ISSUE THREAD

**Issue by Validark** [microsoft/TypeScript#30557]
This fails [on the playground](https://www.typescriptlang.org/play/#src=interface%20TypeA%20%7B%0A%09Name%3A%20%22TypeA%22%3B%0A%09Value1%3A%20%22Cool%20stuff!%22%3B%0A%7D%0A%0Ainterface%20TypeB%20%7B%0A%09Name%3A%20%22TypeB%22%3B%0A%09Value2%3A%200%3B%0A%7D%0A%0Atype%20Type%20%3D%20TypeA%20%7C%20TypeB%3B%0A%0Adeclare%20function%20isType(x%3A%20unknown)%3A%20x%20is%20Type%3B%0A%0Afunction%20WorksProperly(data%3A%20Type)%20%7B%0A%09if%20(data.Name%20%3D%3D%3D%20%22TypeA%22)%20%7B%0A%09%09%2F%2F%20data%3A%20TypeA%0A%09%09const%20value1%20%3D%20data.Value1%3B%0A%09%7D%0A%7D%0A%0Afunction%20DoesNotWork(data%3A%20unknown)%20%7B%0A%09if%20(isType(data))%20%7B%0A%09%09if%20(data.Name%20%3D%3D%3D%20%22TypeA%22)%20%7B%0A%09%09%09%2F%2F%20data%3A%20Type%0A%09%09%09%2F%2F%20data%20should%20be%20TypeA%0A%09%09%09const%20value1%20%3D%20data.Value1%3B%0A%09%09%7D%0A%09%7D%0A%7D%0A), and I believe it also fails on typescript@3.4.0-dev.20190323

**TypeScript Version:**  typescript@3.4.0-dev.20190323

**Search Terms:** unknown intersection failure types narrow

**Code**

```ts
interface TypeA {
	Name: "TypeA";
	Value1: "Cool stuff!";
}

interface TypeB {
	Name: "TypeB";
	Value2: 0;
}

type Type = TypeA | TypeB;

declare function isType(x: unknown): x is Type;

function WorksProperly(data: Type) {
	if (data.Name === "TypeA") {
		// data: TypeA
		const value1 = data.Value1;
	}
}

function DoesNotWork(data: unknown) {
	if (isType(data)) {
		if (data.Name === "TypeA") {
			// data: Type
			// data should be TypeA
			const value1 = data.Value1; // type error!
		}
	}
}
```

**Expected behavior:** in `DoesNotWork`, `data` should be TypeA after we can guarentee it is named "TypeA"

**Actual behavior:** `data` is still only a `Type`

[**Playground Link:** Here](https://www.typescriptlang.org/play/#src=interface%20TypeA%20%7B%0A%09Name%3A%20%22TypeA%22%3B%0A%09Value1%3A%20%22Cool%20stuff!%22%3B%0A%7D%0A%0Ainterface%20TypeB%20%7B%0A%09Name%3A%20%22TypeB%22%3B%0A%09Value2%3A%200%3B%0A%7D%0A%0Atype%20Type%20%3D%20TypeA%20%7C%20TypeB%3B%0A%0Adeclare%20function%20isType(x%3A%20unknown)%3A%20x%20is%20Type%3B%0A%0Afunction%20WorksProperly(data%3A%20Type)%20%7B%0A%09if%20(data.Name%20%3D%3D%3D%20%22TypeA%22)%20%7B%0A%09%09%2F%2F%20data%3A%20TypeA%0A%09%09const%20value1%20%3D%20data.Value1%3B%0A%09%7D%0A%7D%0A%0Afunction%20DoesNotWork(data%3A%20unknown)%20%7B%0A%09if%20(isType(data))%20%7B%0A%09%09if%20(data.Name%20%3D%3D%3D%20%22TypeA%22)%20%7B%0A%09%09%09%2F%2F%20data%3A%20Type%0A%09%09%09%2F%2F%20data%20should%20be%20TypeA%0A%09%09%09const%20value1%20%3D%20data.Value1%3B%0A%09%09%7D%0A%09%7D%0A%7D%0A)

**Related Issues:**  I looked through several pages of issues and did not find anything similar.

**Comment by jack-williams** [comment]
Narrowing by discriminant only applies when the *declared* type of the identifier is a union type; here, the declared type of `data` is `unknown`.

There are probably good reasons for this that someone more knowledgeable can provide, performance possibly among them.

There is an easy workaround which is declaring a new identifier inside the body of the fist condition, then narrowing from that.

**Comment by Validark** [comment]
Personally, I wish this was the way `unknown` worked out of the box because it seems to me that cases like the one I am describing would be extremely common for `unknown` types. The point of `unknown` is that you have to verify what type it is before using it. If you can't use the type normally after checking its type, that is a major problem in my view.

https://github.com/Microsoft/TypeScript/issues/27706 might be related.

**Comment by aleksey-bykov** [comment]
interestingly this works:
```ts
function WorksProperly(data: Type) {
	if (data.Name === "TypeA") {
		// data: TypeA
		const value1 = data.Value1;
	}
}

function DoesNotWork(data: unknown) {
	if (isType(data)) {
		WorksProperly(data);
		// if (data.Name === "TypeA") {
		// 	// data: Type
		// 	// data should be TypeA
		// 	const value1 = data.Value1;
		// }
	}
}
```

**Comment by jack-williams** [comment]
I’m pretty certain the linked issue is not related. The behaviour as I describe is consistent with @aleksey-bykov ‘s example. The declared argument type of `DoesNotWork` is not a union, while the declared argument type of `WorksProperly` is.

While it is nice to have full reasoning applied to every situation, TypeScript gains a lot of performance wins by judiciously picking when and where to apply expensive checks.

I think this might be one of those cases, but I’m not certain. Someone on the team will be able to correct me.

**Comment by aleksey-bykov** [comment]
what i've been told one day here is that narrowing doesn't **make new types** it only **narrows** what's given, in case of `unknown` it is implied all the types that only can be imagined out there (over 9000), which is well... 
in case of narrowing from `unknown` the design team clearly took a short cut, and here we see the consequences

**Comment by RyanCavanaugh** [comment]
This feels like it might be a bug. Once `data` is narrowed to a union type, narrowing by the discriminant is entirely reasonable

**Comment by jack-williams** [comment]
Fix here #30593

**Comment by talbenari1** [comment]
Out of curiosity, is this related? I didn't have to use `unknown` to lose type refinement.

```typescript
type Foo = Bar | Baz;

interface Bar {
    kind: 'Bar';
    prop: string | void;
}


interface Baz {
    kind: 'Baz';
    prop: string;
}

function fn(el: Foo): number {
    switch (el.kind) {
        case 'Bar':
            if (!el.prop) {
                return 0;
            }
        // falls through
        case 'Baz':
            return el.prop.length; // Property 'length' does not exist on type 'string | void'
    }
}
```
